### PR TITLE
xds: respond to requests previously miscategorized as ACKs

### LINKED ISSUE #38709: Proxy startup stalled by warming EDS clusters
https://prow.istio.io/view/gs/istio-prow/logs/integ-k8s-124_istio_postsubmit/1521489723254640640

What we see is version 6 is pushed, but version 9 is stuck warming. For ingress gw, it startsup since it completed with v6, but for egress v9 must have blocked the init.

```
2022-05-03T14:27:08.433846Z     info    ads     ADS: new connection for node:istio-egressgateway-56cf788c86-x2x58.istio-system-2
2022-05-03T14:27:08.434802Z     info    ads     CDS: PUSH request for node:istio-egressgateway-56cf788c86-x2x58.istio-system resources:18 size:18.1kB cached:0/17
2022-05-03T14:27:08.441481Z     info    ads     Push debounce stable[8] 10 for config EnvoyFilter/istio-system/stats-filter-1.15 and 9 more configs: 100.201218ms since last change, 114.869484ms since last push, full=true
2022-05-03T14:27:08.446618Z     info    ads     XDS: Pushing:2022-05-03T14:27:08Z/7 Services:6 ConnectedEndpoints:2 Version:2022-05-03T14:27:08Z/7
2022-05-03T14:27:08.447737Z     info    ads     CDS: PUSH for node:istio-egressgateway-56cf788c86-x2x58.istio-system resources:18 size:18.1kB cached:0/17
2022-05-03T14:27:08.448289Z     info    ads     EDS: PUSH request for node:istio-egressgateway-56cf788c86-x2x58.istio-system resources:17 size:2.0kB empty:1 cached:15/17
2022-05-03T14:27:09.036214Z     info    ads     Push debounce stable[9] 10 for config EnvoyFilter/istio-system/tcp-stats-filter-1.12 and 9 more configs: 100.339244ms since last change, 109.692538ms since last push, full=true
2022-05-03T14:27:09.041912Z     info    ads     XDS: Pushing:2022-05-03T14:27:09Z/8 Services:6 ConnectedEndpoints:2 Version:2022-05-03T14:27:09Z/8
2022-05-03T14:27:09.042957Z     info    ads     CDS: PUSH for node:istio-egressgateway-56cf788c86-x2x58.istio-system resources:18 size:18.1kB cached:10/17
2022-05-03T14:27:09.043164Z     info    ads     EDS: PUSH for node:istio-egressgateway-56cf788c86-x2x58.istio-system resources:17 size:2.0kB empty:0 cached:17/17
2022-05-03T14:27:09.274697Z     info    ads     XDS: Pushing:2022-05-03T14:27:09Z/9 Services:6 ConnectedEndpoints:2 Version:2022-05-03T14:27:09Z/9
2022-05-03T14:27:09.275656Z     info    ads     CDS: PUSH for node:istio-egressgateway-56cf788c86-x2x58.istio-system resources:18 size:18.3kB cached:8/17
```

No EDS request. Either envoy bug or Istio filters the request

### PR DESCRIPTION
**Pull request by howardjohn**
See https://github.com/envoyproxy/envoy/issues/13009 for details

Fixes #38709 (previously 'fixed', but really the fix was a workaround)

Fixes https://github.com/istio/istio/issues/39720

**Please provide a description of this PR:**

### CODE REVIEW

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -427,6 +428,17 @@ func (s *DiscoveryServer) shouldRespond(con *Connection, request *discovery.Disc
 	// Envoy can send two DiscoveryRequests with same version and nonce
 	// when it detects a new resource. We should respond if they change.
 	if len(removed) == 0 && len(added) == 0 {
+		// We got a request which looks like an ACK, but we already got the same ACK. Envoy won't ACK the
+		// same resource twice. However, in some cases a request for new resources is indistinguishable
+		// from an ACK (in particular, when requesting EDS after a CDS push):
+		// https://github.com/envoyproxy/envoy/issues/13009. As a result, if we see the same ACK multiple
+		// times, we treat subsequent requests as something we should respond to.
+		log.Errorf("howardjohn: %v vs %v", request.ResponseNonce, previousNonceAcked)
```
Comment (nmittler) [inline pilot/pkg/xds/ads.go:436]: Should we remove your name? :)

**Review by nmittler — APPROVED** [review]

**Comment by ramaraochavali** [comment]
LGTM.

File: pilot/pkg/xds/ads.go [hunk 2]
```diff
@@ -413,6 +413,7 @@ func (s *DiscoveryServer) shouldRespond(con *Connection, request *discovery.Disc
 
 	// If it comes here, that means nonce match. This an ACK. We should record
 	// the ack details and respond if there is a change in resource names.
+	previousNonceAcked := previousInfo.NonceAcked
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/ads.go:416]: Comments need to be updated

**Review by hzxuzhonghu — COMMENTED** [review]
Is there a plan when envoy would address the ack and request using same nonce?

**Comment by howardjohn** [comment]
@hzxuzhonghu there is some discussion in https://github.com/envoyproxy/envoy/issues/13009, doesn't seem soon

**Pull request merged**